model/task/task.go

```diff
@@ -871,35 +884,240 @@ func (t *Task) ActivateTask(caller string) error {
 	t.ActivatedBy = caller
 	t.Activated = true
 	t.ActivatedTime = time.Now()
-	return UpdateOne(
+
+	return ActivateTasks([]Task{*t}, caller)
+}
+
+func ActivateTasks(tasks []Task, caller string) error {
+	taskIDs := make([]string, 0, len(tasks))
+	for _, t := range tasks {
+		taskIDs = append(taskIDs, t.Id)
+	}
+
+	_, err := UpdateAll(
 		bson.M{
-			IdKey: t.Id,
+			IdKey: bson.M{"$in": taskIDs},
 		},
 		bson.M{
 			"$set": bson.M{
 				ActivatedKey:     true,
 				ActivatedByKey:   caller,
-				ActivatedTimeKey: t.ActivatedTime,
+				ActivatedTimeKey: time.Now(),
 			},
 		})
+	if err != nil {
+		return errors.Wrap(err, "can't activate task")
+	}
+
+	for _, t := range tasks {
+		event.LogTaskActivated(t.Id, t.Execution, caller)
+	}
+
+	return ActivateDeactivatedDependencies(taskIDs, caller)
+}
+
+// ActivateDeactivatedDependencies activates tasks that depend on these tasks which were deactivated because a task
+// they depended on was deactivated. Only activate when all their dependencies are activated or are being activated
+func ActivateDeactivatedDependencies(tasks []string, caller string) error {
+	taskMap := make(map[string]bool)
+	for _, t := range tasks {
+		taskMap[t] = true
+	}
+
+	depTaskMap := make(map[string]bool)
+	tasksDependingOnTheseTasks, err := getRecursiveDependenciesDown(tasks, nil)
+	if err != nil {
+		return errors.Wrap(err, "can't get recursive dependencies down")
+	}
+	for _, t := range tasksDependingOnTheseTasks {
+		depTaskMap[t.Id] = true
+	}
+
+	// get dependencies we don't have yet and add them to a map
+	tasksToGet := []string{}
+	for _, t := range tasksDependingOnTheseTasks {
+		if t.Activated || !t.DeactivatedForDependency {
+			continue
+		}
+
+		for _, dep := range t.DependsOn {
+			if !taskMap[dep.TaskId] && !depTaskMap[dep.TaskId] {
+				tasksToGet = append(tasksToGet, dep.TaskId)
+			}
+		}
+	}
+	missingTasks, err := FindAll(db.Query(bson.M{IdKey: bson.M{"$in": tasksToGet}}).WithFields(ActivatedKey))
+	if err != nil {
+		return errors.Wrap(err, "can't get missing tasks")
+	}
+	missingTaskMap := make(map[string]Task)
+	for _, t := range missingTasks {
+		missingTaskMap[t.Id] = t
+	}
+
+	// do a topological sort so we've adjudicated all a task's dependencies by the time
+	// we get up to it
+	sortedDependencies, err := topologicalSort(tasksDependingOnTheseTasks)
+	if err != nil {
+		return errors.Wrap(err, "can't do topological sort")
+	}
+	tasksToActivate := make(map[string]Task)
+	for _, t := range sortedDependencies {
+		if t.Activated || !t.DeactivatedForDependency {
+			continue
+		}
+
+		depsSatisfied := true
+		for _, dep := range t.DependsOn {
+			// not being activated now
+			if _, ok := tasksToActivate[dep.TaskId]; !ok && !taskMap[dep.TaskId] {
+				// and not already activated
+				if depTask := missingTaskMap[dep.TaskId]; !depTask.Activated {
```

Comment:
Should we also check ok here?

Reply:
I thought the effect of having it would be the same as leaving it out (because `depTask.Activated` will be false when `!ok`)